### ISSUE TITLE
Explore AutoFill Credential Provider extension

### ISSUE THREAD

**Issue by devinreams** [mozilla-lockbox/lockbox-ios#476]
https://developer.apple.com/videos/play/wwdc2018/721

**Comment by joeyg** [comment]
So exciting! Here are the docs https://developer.apple.com/documentation/authenticationservices/ascredentialidentitystorestate?language=objc

**Comment by devinreams** [comment]
@sashei thanks for the nice summary write-up, this is probably a good place for us to start discussing and deciding what we want to build and who else we want to coordinate with (iOS mobile team).

@mozilla-lockbox/product we should consider a mini project kick-off or discussion to talk through this together and make sure everyone "gets" what we can do here. I'll set up a meeting for next week...

_FWIW I dropped in some screenshots I captured while watching the video up above because I love visuals_ ☝️

**Comment by devinreams** [comment]
Notes from June 20 team meeting:

## Overview

- autofill for native applications _that have implemented the specific username/password fields provided by Apple_, previously could only be filled with keychain credentials
  - also available to Safari for form fields
  - credentials will appear in the **QuickType bar** above the keyboard
  - if they don't see what they want, an **Entry List** can be displayed, best practice says to allow showing all items (but filter those that relate to the app/domain being requested)

- all of this matches minimal login model since we only get the username back (then iOS provides just that one password)
  - if we want to lock/unlock, such as providing a "master password", can display an **minimal  authentication prompt** before that credential is returned (we should match the state our app is in: unlocked, Face ID)

- takes shape in an extension, kinda a separate mini app in its own container
  -  Apple expects us to maintain data in `ASCredentialIdentityStore` separate from our app ("Best practices: Keep the credential identity store up-to-date")
  - we can provide some metadata to help "rank" order the items shown

- we can allow users to configure in **settings** and less clear what is required vs. optional
  - at very least just "enable" us as a provider, but not sure what states things are in (does it know if we've even authenticated?)

  - we can "compete" with other credential providers since multiple could be enabled, unclear what that will look like

## Quick List of Things to Build

This will be an epic with consolidated design/user stuff (that we can split out as needed to engineering stories).

- [ ] Project / Autofill Credential Provider configuration
- [ ] `ASCredentialIdentityStore` to store and update data
- [ ] List view
  - no need for detail entry view
  - high potential for reuse, minimal stripped-away design
- [ ] Minimal authentication view (to unlock with Face ID / Touch ID / passcode)
- [ ] User onboarding to explain autofill during first run of Lockbox app
- [ ] Telemetry events
- [ ] More content in FAQ
- [ ] Link from our settings screen to get to autofill settings

---

From here I will spin up a new issue and ask @mozilla-lockbox/product to fill in the blanks around user stories and acceptance criteria and designs.

In the meantime, @sashei can you please help explore the Settings stuff a bit more?

We will also chat with the other Mozilla mobile folks about their interest and ability in collaborating here (consult and discuss? get help with code review? collaborate and write code together?)

**Comment by st3fan** [comment]
I can't figure out if the "Standard "extension" single view screen" is required. Is this an optional part? Can you just to the suggestion in the QuickType bar?

And I also wonder what the security story is. Will Apple ask for TouchID/FaceID before even suggesting credentials through an extension or does the extension have to implement that?

**Comment by devinreams** [comment]
>I can't figure out if the "Standard "extension" single view screen" is required. Is this an optional part? Can you just do the suggestion in the QuickType bar? (Which may be a good start)

Yeah it seems like that's the screen they're suggesting we need for when they hit the "key icon" 🔑 to see all passwords. I wondered too if they just give us the system list of logins but I don't think so. But I also don't know how multiple password providers would be handled (prompt to open 1Password list vs iCloud Keychain list vs Lockbox list?!). To be determined...

>And I also wonder what the security story is. Will Apple ask for TouchID/FaceID before even suggesting credentials through an extension or does the extension have to implement that?

We're fairly confident the story is: the app extension gets to pick what (If any) security we put in place before the password is delivered to the app and not to assume any other system-level security going up in front. To be confirmed...

**Comment by nickbrandt** [comment]
> But I also don't know how multiple password providers would be handled (prompt to open 1Password list vs iCloud Keychain list vs Lockbox list?!). To be determined...

I saw this video on Keeper's site, suggesting an Action Sheet would be presented in that case.

https://keepersecurity.com/blog/2018/06/06/awesome-news-keeperfill-ios-12-seamless-password-autofill/
